lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java

```
@@ -166,25 +173,35 @@ public static NeighborQueue search(
   /**
    * Returns the {@link NeighborQueue} connected to the given node.
    *
+   * @param level level of the graph
    * @param node the node whose neighbors are returned
    */
-  public NeighborArray getNeighbors(int node) {
-    return graph.get(node);
+  public NeighborArray getNeighbors(int level, int node) {
+    NeighborArray result = graph.get(level).get(node);
+    assert result != null;
+    return result;
   }
 
   @Override
   public int size() {
-    return graph.size();
+    return graph.get(0).size(); // all nodes are located on the 0th level
   }
 
-  int addNode() {
-    graph.add(new NeighborArray(maxConn + 1));
-    return graph.size() - 1;
+  public void addNode(int level, int node) {
+    if (level > 0) {
+      // Levels above 0th don't contain all nodes,
+      // so for missing nodes we add null NeighborArray
+      int nullsToAdd = node - graph.get(level).size();
+      for (int i = 0; i < nullsToAdd; i++) {
```

This seems kind of expensive. Perhaps we should have a data structure that has NeighborArray, with values that are dense ordinals in the current layer *plus* the ordinal of the current node in the next layer. If we do that, each layer can be dense instead of creating N*numlayers entries, most of which are null. Maybe just throw an UnsupportedOperationException here for now and we can resolve this as we get deeper into the implementation though

---

OK I realize now that of course we need to encode the (layer 0) node ids in each other layer in order to be able to look up vector values. We can optimize RAM usage later.

---

Indeed, this is not the ideal implementation, as we would need extra RAM to store all references to null objects. 
I've added a TODO to optimize RAM later.  
Or would you suggest to do this optimization in this PR? We can keep only dense ordinals in the current layer, and add another field to `HnswGraph` : `List<List<Integer>> nodesByLevels` that will encode the (layer 0) node ids in each other layers.

---

Right, I think it's OK to commit this and follow up with an improvement. When we do it, let's try to use primitive arrays `int[]` rather than `List<Integer>` to avoid autoboxing. Binary search seems like a good choice (rather than another data structure for direct lookup) given we only do this numLevels times for each search.